js.go

```diff
@@ -2751,10 +2753,18 @@ func (sub *Subscription) Fetch(batch int, opts ...PullOpt) ([]*Msg, error) {
 					// wait this time.
 					noWait = false
 					err = sendReq()
-				} else if err == ErrTimeout && len(msgs) == 0 {
+				} else if err == ErrTimeout {
 					// If we get a 408, we will bail if we already collected some
```

Comment:
@wallyqs I'm not sure that's needed after this change, but I may be wrong. Was this added to suppress timeout from previous `Fetch()`?

Reply:
This was to avoid a client getting stuck on 408 errors due to reaching the max waiting limits, I think behavior of the server has advanced enough that this might not be longer needed as you mention https://github.com/nats-io/nats.go/commit/76690683f2402cac2c132034611debd660ead77e

There was a test that got added for that but we removed it at some point, so maybe could add a version of it back just in case: https://github.com/nats-io/nats.go/commit/76690683f2402cac2c132034611debd660ead77e#diff-9e6bdee4051a6e36d6d507dcef63b5f9a39b281e421c08f02ec74a0fb6639e3aR4501

Reply:
I think we should leave this as it is.

Test you mentioned is failing with the newer versions of the server, but the behavior of Fetch is different than FetchBatch, as it operates with both `no_wait` true & false.

Let's leave it here as it is.

Reply:
yes looks like would be better to leave it as is for now, it might still possible to run into multiple 408s condition according to that test when calling it concurrently:

```
=== RUN   TestJetStream_PullSubscribeMaxWaiting/psub_n=3_r=1/blocking_fetch
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUDx, err=nats: requests pending, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[99] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUAf, err=<nil>, msgs=0, header=map[]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUHF, err=nats: requests pending, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[99] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUKX, err=<nil>, msgs=0, header=map[]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUDx, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[99] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUNp, err=<nil>, msgs=0, header=map[]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUHF, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[99] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUAf, err=<nil>, msgs=1, header=map[]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUDx, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[99] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUKX, err=<nil>, msgs=1, header=map[]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUNp, err=nats: requests pending, msgs=1, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUHF, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUAf, err=nats: requests pending, msgs=2, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUDx, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUKX, err=nats: requests pending, msgs=2, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUHF, err=<nil>, msgs=0, header=map[]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUDx, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUHF, err=nats: timeout, msgs=1, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUDx, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUHF, err=nats: timeout, msgs=1, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUDx, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUHF, err=nats: timeout, msgs=1, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
inbox=_INBOX.6c2FnkK9m5H9HgXWpzUT6r.6c2FnkK9m5H9HgXWpzUUDx, err=nats: timeout, msgs=0, header=map[Description:[Request Timeout] Nats-Pending-Bytes:[0] Nats-Pending-Messages:[1] Status:[408]]
```